### PR TITLE
Fix issues with package dependencies.

### PR DESCRIPTION
**Pull request by stuartarchibald**
* Removes the unconditional installation of ipykernel (the
  "VANILLA_INSTALL" must have no additional packages).
* Fixes the package requirements, ipython_genutils is a dependency
  of ipykernel so does not need listing separately.
* Updates the associated comments.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Note:
```
$ conda search ipykernel --info
```
Sample output:
```
ipykernel 6.4.1 py39h06a4308_1
------------------------------
file name   : ipykernel-6.4.1-py39h06a4308_1.conda
name        : ipykernel
version     : 6.4.1
build       : py39h06a4308_1
build number: 1
size        : 194 KB
license     : BSD-3-Clause
subdir      : linux-64
url         : https://repo.anaconda.com/pkgs/main/linux-64/ipykernel-6.4.1-py39h06a4308_1.conda
md5         : bfb323ce72a0787939d2412529a8063f
timestamp   : 2021-10-06 15:38:18 UTC
dependencies: 
  - debugpy >=1,<2.0
  - ipython >=7.23.1,<8.0
  - ipython_genutils
  - jupyter_client <8.0
  - matplotlib-inline >=0.1.0,<0.2.0
  - python >=3.9,<3.10.0a0
  - tornado >=4.2,<7.0
  - traitlets >=4.1.0,<6.0

```

**Comment by stuartarchibald** [comment]
xref #7345 for the new tests that introduced this dependency.

File: buildscripts/incremental/setup_conda_environment.sh [hunk 1]
```diff
@@ -53,9 +53,11 @@ set -v
...
 # Install optional packages into activated env
 if [ "${VANILLA_INSTALL}" != "yes" ]; then
-    # Scipy, CFFI, jinja2, IPython, ipython_genutils and pygments are optional dependencies, but exercised in the test suite
-    # pexpect is used to run the
-    $CONDA_INSTALL ${EXTRA_CHANNELS} cffi jinja2 ipython ipython_genutils pygments pexpect
+    # Scipy, CFFI, jinja2, IPython, and pygments are optional dependencies,
+    # but exercised in the test suite.
+    # pexpect is used to run the gdb tests.
+    # ipykernel is used for testing ipython behaviours.
+    $CONDA_INSTALL ${EXTRA_CHANNELS} cffi jinja2 ipython pygments pexpect ipykernel
```
Comment (esc) [inline buildscripts/incremental/setup_conda_environment.sh:60]:
```suggestion
    $CONDA_INSTALL ${EXTRA_CHANNELS} cffi jinja2 ipython ipython_genutils pygments pexpect ipykernel
```

**Review by esc — CHANGES_REQUESTED** [review]
I think that `ipython_genutils` is missing from the install command.

I would also suggest to add `ipython_genutils` and `ipykernel` to the dependency list at:

https://numba.readthedocs.io/en/stable/user/installing.html#dependency-list

Tracking optional dependencies in CI scripts is somewhat suboptimal.

**Comment by esc** [comment]
> > I think that `ipython_genutils` is missing from the install command.
> 
> Note: [#7626 (comment)](https://github.com/numba/numba/pull/7626#issuecomment-987783977), `ipython_genutils` is a hard dependency of `ipykernel`.

Oh yes, I missed that.

> > I would also suggest to add `ipython_genutils` and `ipykernel` to the dependency list at:
> > https://numba.readthedocs.io/en/stable/user/installing.html#dependency-list
> > Tracking optional dependencies in CI scripts is somewhat suboptimal.
> 
> I don't think these are optional dependencies for Numba, nothing is gained from a user's perspective by installing them and the Numba compiler itself does not use them? These packages are used solely in the test suite to check various features. Another example is `pexpect`, the test suite needs this to drive the `gdb` integration testing, but Numba has no reliance on it, optional or otherwise. The conda build `meta.yaml` test section reflects this. If anything I'd suggest the docs might need an "Optional test time" section?

Makes sense. Probably not important for now.

**Review by esc — DISMISSED** [review]

**Comment by esc** [comment]
@stuartarchibald thank you for the patches!

**Review by sklam — APPROVED** [review]
LGTM

**Pull request merged**